an-introduction-to-unit-testing.md

```diff
@@ -0,0 +1,852 @@
+
+Example:  `OpticalSensor_IsRedBloodCell_ReturnsFalse()`
+
+### Structure of a unit test
+
+The typical sequence of actions performed by a unit test is the following: creating an object, setting it up as necessary, acting on the object, and asserting that something is as expected.
+
+### Attributes
+
+The unit testing framework provides several attributes used to decorate the code of the unit tests, such as  `TestFixture`,  `Test`,  `TestCase`,  `Setup`, and  `TearDown`.
+
+### Assertions
+
+The unit testing also provides classes and methods to assert what is expected, such as  `Assert.IsFalse()`,  `Assert.IsTrue()`,  `Assert.AreEqual()`,  `Assert.Throws()`, and  `Assert.DoesNotThrow()`.
+
+## Fakes
+
+This section presents faking techniques commonly used in unit testing.
+
+### External dependencies
+
+An external dependency is an object in your system that your code under test interacts with and over which you have no control.
+
+### Fakes: stubs and mocks
+
+A fake is a generic term that can describe either a stub or a mock. A stub is a controllable replacement for an existing dependency: by using a stub, you can test your code without dealing with the dependency directly. Mocks are just like stubs, but you can assert against the mock object, whereas you do not assert against a stub.
+
+### Abstraction and dependency injection
+
+When you cannot test something, either you add a layer that wraps up the calls to that something, and then mimic that layer in your tests, or you make that something replaceable, so that it is itself a layer of indirection.
+
+### Refactoring
+
+Refactoring means changing a piece of code without changing its functionality. There are two types of dependency-breaking refactoring, and one depends on the other: abstracting concrete objects into interfaces, and refactoring to allow the injection of fakes of those interfaces. In order to abstract concrete objects into interfaces, extract an interface to allow replacing the underlying implementation. In order to allow the injection of fakes of the interfaces, inject a stub into a class under test, inject a mock using a constructor, inject a mock using a property, or inject a mock using a method.
+
+### Isolation frameworks
+
+An isolation framework is a set of programmable APIs that makes the creation of fake objects much faster and simpler. Isolation frameworks exist for most languages that have a unit testing framework associated with them.
+
+## Code samples
+
+This section describes the examples included in my unit testing training project available at the following location.
+
+[https://github.com/fabioscagliola/Unit-Testing](https://github.com/fabioscagliola/Unit-Testing)
+
+### Unit testing example 1
+
+The **UnitTesting01.cs** file includes two unit tests that verify the Boolean values returned by the  `IsRedBloodCell`  method of the  `OpticalSensor`  class.
+
+```csharp
```

Comment:
Just a suggestion: Write the file name with the inline code style as well. e.g. `UnitTest01.cs`